### ISSUE TITLE
Implement file system module

### ISSUE THREAD

**Issue by geckon** [qorelanguage/qore#2000]
I plan to add a module providing support for file system related operations.

Possible features:
- Path oprations like join, split, parent_dir, is_absolute, normalize, is_equal,... These will probably mostly be string manipulations (Path can also be implemented as a class providing validation etc. if we find that useful) that don't need to actually touch the file system.
- "Real" file system operations - copy, move, remove, list, find,...
- Temporary files/dirs handling functions

The module should be multiplatform (POSIX and Windows) so programmers don't need to know/care about the system their code will run on. It can be achieved either by platform-specific modules (the correct one is loaded in run-time depending on the actual platform which is how e.g. Python does it as far as I know) or by classes implementing a common interface (some code could be shared that way).

I know that some of the proposed functions are already implemented in Qore itself and its Util module - there we can consider whether it makes sense to move the functionality to the new module or not.

I'm open to ideas and suggestions.

**Comment by geckon** [comment]
Please don't close this issue once #2976 is merged unless we decide to open new issues for the following ideas and improvements:
- ~~more user friendly interface for temporary files/dirs should be added - i.e. no need for removing the temp files after using them etc.~~ (see #2999)
- add "overwrite" parameter to `copy_path()` and `copy_tree()`?
- ~~add "fail_immediately" parameter to `remove_tree()` and `copy_tree()`~~ (done in a770846 and 04f75a2)
- add "ignore_errors" parameter to `remove_tree()` and `copy_tree()`?
- ~~something like Python [walk()](https://docs.python.org/3/library/os.html#os.walk) and [glob()](https://docs.python.org/3.7/library/glob.html)~~ (see #3000)
- add "follow_symlinks" parameter to remove functions?
- add `move_*()` wrappers for `rename()` function to align with `copy_*()` and `remove_*()`?


Feel free to throw more ideas in and/or comment on these. We can either track them here or open a separate issue for each valid idea selected for development.

**Comment by geckon** [comment]
Since #2976 has been merged, I propose to close this issue and add new issues for these if we want to see them in FsUtil:
- add "overwrite" parameter to `copy_path()` and `copy_tree()`?
- add "ignore_errors" parameter to `remove_tree()` and `copy_tree()`?
- add "follow_symlinks" parameter to remove functions?
- add `move_*()` wrappers for `rename()` function to align with `copy_*()` and `remove_*()`?

I already added issues for other two improvements I think we should do: #2999 and #3000.
